### DISCUSSION TITLE
how to use kernel_libipsec

### DISCUSSION THREAD

**Discussion by songxingping** [strongswan/strongswan#2338] [Q&A] [answered]
on ubuntu 18.04, we want to use kernel_libipsec on userland instead of xfrm, we configure with --diseable-kernel-netlink --enable-kernel-libipsec --enable-kernel-pfroute. then we got this "not found <net/if_dl.h>" or "cannot complie this plugin on system where 'struct sockaddr' has no sa_len member".
how can i get continue?

**Comment by tobiasbrunner** [comment]
> on ubuntu 18.04, we want to use kernel_libipsec on userland instead of xfrm,

Why? Did you read [the docs](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html)?

> --enable-kernel-pfroute

This plugin only works on FreeBSD/macOS. You'll still need the _kernel-netlink_ plugin on Linux even when you don't want to use XFRM.

**Reply by songxingping** [reply]
thank you for your answer.
i have read this doc times, just 'On systems that use the kernel-pfroute plugin ([FreeBSD] or [macOS]) a separate TUN device will be created for each [virtual IP](https://docs.strongswan.org/docs/5.9/features/vip.html). On Linux this is not required' relates to linux and libipsec.
does 'not required' means not support?

**Reply by tobiasbrunner** [reply]
No, on Linux a single TUN device can be used, which is created right when the plugin is loaded. The other platforms require separate TUN devices for virtual IPs even when not using the _kernel-netlink_ plugin in order to force the virtual IP as the source address for local packets (on Linux this can be achieved via preferred source IP on the routes).

**Reply by songxingping** [reply]
so if i want to let encrypt or decrypt on userland using myself algorithm,  kernel-libipsec and kernel-netlink should been load together, and make sure kernel-libipsec load before kernel-netlink. charon will use libipsec 's kernel-ipsec feature and kernel-netlink's kernel-net feature, is right?

**Reply by tobiasbrunner** [reply, marked as answer]
Yes, exactly (that should be the default order if both plugins are enabled, by the way).

**Reply by songxingping** [reply]
thank you so much
